From: James McDonald <james@jamesmcdonald.com>
Description: Add an additional stat() check if a d_type check fails for
 kbfunc_exec lookups. The d_type check doesn't work on all filesystems,
 eg XFS.
--- a/kbfunc.c
+++ b/kbfunc.c
@@ -246,6 +246,7 @@
 	struct menu		*mi;
 	struct menu_q		 menuq;
 	int			 l, i, cmd = arg->i;
+	struct stat		sb;
 
 	switch (cmd) {
 	case CWM_EXEC_PROGRAM:
@@ -276,14 +277,19 @@
 			continue;
 
 		while ((dp = readdir(dirp)) != NULL) {
-			/* skip everything but regular files and symlinks */
-			if (dp->d_type != DT_REG && dp->d_type != DT_LNK)
-				continue;
 			(void)memset(tpath, '\0', sizeof(tpath));
 			l = snprintf(tpath, sizeof(tpath), "%s/%s", paths[i],
 			    dp->d_name);
 			if (l == -1 || l >= sizeof(tpath))
 				continue;
+			/* skip everything but regular files and symlinks */
+			if (dp->d_type != DT_REG && dp->d_type != DT_LNK) {
+				/* use an additional stat-based check in case d_type isn't supported */
+				if (lstat(tpath, &sb) < 0)
+					continue;
+				if (!S_ISREG(sb.st_mode) && !S_ISLNK(sb.st_mode))
+					continue;
+			}
 			if (access(tpath, X_OK) == 0)
 				menuq_add(&menuq, NULL, "%s", dp->d_name);
 		}
